### PR TITLE
[Merged by Bors] - feat: Stabilizers of sets/finsets

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove a bunch of results about `stabilizer G s` where `Group G`, `MulAction G α`, `s : Set α`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Pointwise/Stabilizer.lean [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+@[to_additive (attr := simp)]
+lemma stabilizer_singleton (b : α) : stabilizer G ({b} : Set α) = stabilizer G b := by ext; simp
+
+@[to_additive]
+lemma mem_stabilizer_set {s : Set α} : a ∈ stabilizer G s ↔ ∀ b, a • b ∈ s ↔ b ∈ s := by
+  refine ⟨fun h b ↦ ?_, fun h ↦ ?_⟩
+  · rw [← (smul_mem_smul_set_iff : a • b ∈ a • s ↔ _), mem_stabilizer_iff.1 h]
+  simp_rw [mem_stabilizer_iff, Set.ext_iff, mem_smul_set_iff_inv_smul_mem]
+  exact ((MulAction.toPerm a).forall_congr' $ by simp [Iff.comm]).1 h
```
Comment (eric-wieser) [inline Mathlib/Algebra/Pointwise/Stabilizer.lean:35]: I imagine rewriting by `mem_stabilizer_iff` before the case split avoids having to rewrite twice.

File: Mathlib/Algebra/Pointwise/Stabilizer.lean [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+@[to_additive (attr := simp)]
+lemma stabilizer_mul_self (s : Set G) : (stabilizer G s : Set G) * s = s := by
+  ext
+  refine ⟨?_, fun h ↦ ⟨_, _, (stabilizer G s).one_mem, h, one_mul _⟩⟩
+  rintro ⟨a, b, ha, hb, rfl⟩
+  rw [← mem_stabilizer_iff.1 ha]
+  exact smul_mem_smul_set hb
+
+end Set
```
Comment (eric-wieser) [inline Mathlib/Algebra/Pointwise/Stabilizer.lean:59]: Can you add some subheadings for `Set` and `Finset`?

File: Mathlib/GroupTheory/GroupAction/Basic.lean [hunk 3]
```diff
@@ -457,6 +453,29 @@ theorem mem_stabilizer_iff {a : α} {g : G} : g ∈ stabilizer G a ↔ g • a =
...
+  simp_rw [SetLike.le_def, mem_stabilizer_iff, ← smul_assoc]; rintro a h; rw [h]
+
+@[to_additive]
+lemma le_stabilizer_smul_right [SMul α β] [SMulCommClass G α β] (a : α) (b : β) :
+    stabilizer G b ≤ stabilizer G (a • b) := by
+  simp_rw [SetLike.le_def, mem_stabilizer_iff, smul_comm]; rintro a h; rw [h]
+
+@[to_additive (attr := simp)]
+lemma stabilizer_mul_eq_left [Group α] [IsScalarTower G α α] (a b : α) :
+    stabilizer G (a * b) = stabilizer G a := by
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/GroupAction/Basic.lean:468]: Does this generalize to a `smul`?

File: Mathlib/Algebra/Pointwise/Stabilizer.lean (hunk 2 above)
Comment (YaelDillies) [reply Mathlib/Algebra/Pointwise/Stabilizer.lean:59]: Sure. Btw I really dislike how I have to `MulAction.something_stabilizer_set`/`MulAction.something_stabilizer_finset`. I would much prefer being able to write `Set.something_mulStabilizer`/`Finset.something_addStabilizer`. Do you think it's worth a rename?

File: Mathlib/Algebra/Pointwise/Stabilizer.lean [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+@[to_additive (attr := simp)]
+lemma stabilizer_subgroup (s : Subgroup G) : stabilizer G (s : Set G) = s := by
+  simp_rw [SetLike.ext_iff, mem_stabilizer_set]
+  refine fun a ↦ ⟨fun h ↦ ?_, fun ha b ↦ s.mul_mem_cancel_left ha⟩
+  simpa only [smul_eq_mul, SetLike.mem_coe, mul_one] using (h 1).2 s.one_mem
+
+@[to_additive]
+lemma map_stabilizer_le (f : G →* G) (s : Set G) :
+    (stabilizer G s).map f ≤ stabilizer G (f '' s) := by
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Pointwise/Stabilizer.lean:45]: Yep, verbatim!

File: Mathlib/Algebra/Pointwise/Stabilizer.lean [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+@[to_additive]
+lemma mem_stabilizer_set {s : Set α} : a ∈ stabilizer G s ↔ ∀ b, a • b ∈ s ↔ b ∈ s := by
+  refine ⟨fun h b ↦ ?_, fun h ↦ ?_⟩
+  · rw [← (smul_mem_smul_set_iff : a • b ∈ a • s ↔ _), mem_stabilizer_iff.1 h]
+  simp_rw [mem_stabilizer_iff, Set.ext_iff, mem_smul_set_iff_inv_smul_mem]
+  exact ((MulAction.toPerm a).forall_congr' $ by simp [Iff.comm]).1 h
+
+@[to_additive (attr := simp)]
+lemma stabilizer_subgroup (s : Subgroup G) : stabilizer G (s : Set G) = s := by
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Pointwise/Stabilizer.lean:38]: There are two `MulOpposite` statements actually, which makes the situation even more irritating given that I can't seem to generalise the statement to two groups. `stabilizer G _ : Subgroup G`, so the only option would be to generalise with a map on one side, but then I need that map to be injective and equivariant or something like that. I certainly can't get much more general than action of a subgroup on a group.

File: Mathlib/GroupTheory/GroupAction/Basic.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/GroupTheory/GroupAction/Basic.lean:468]: Yes, but with an `Injective (· • b : α → β)` that's not covered by any typeclass.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**